### PR TITLE
Login credential store

### LINKED ISSUE #2732: Stay logged in Festure not working in NC 11
After reading https://github.com/nextcloud/server/issues/1067 i thought this Problem is solved with NC 11, but it is not.

### Steps to reproduce
1.  Consider "External Storage" is disabled
2. Logout from NC > You are back at the Login Screen
3. A Checkbox is available to "Stay logged in"
4. Loggin
5. Close the  Browser
6. Reopen the Browser and Open NC URL
7. Your are logged in
8. Enable "External Storage" App
9. Close Browser
10. Reopen Browser with NC URL
11. You are logged out and at the Login Mask there is no "Stay logged  in" Checkbox

### Expected behaviour
The "Stay logged in" Feature should be available independent if i have "External Storage" enabled or not. 

### Actual behaviour
"Stay Logged in" is not available, if "External Storage" is enabled

### Server configuration
**Operating system**: Raspbian (Debian Jessie)

**Web server:** Nginx 1.9

**Database:** Maria DB 10

**PHP version:** 7.0

**Nextcloud version:** NC 11

**Are you using external storage, if yes which one:** local

**Are you using encryption:** no

### PR DESCRIPTION
**Pull request by ChristophWurst**
Since NC 10 we use session tokens for every connected browser and allow devices to use an app-specific password. Those tokens store the encrypted login password, which we can now use to authenticate external storages. In addition, this functionality is useful for other apps like Mail, where we want to create an account automatically, see https://github.com/nextcloud/mail/pull/28.
This way we can allow remember-me logins for all users regardless of enabled apps. Until now, remember-me was disabled as soon as external storage support was enabled because the login password was stored in the session.

TODO:
- [x] ~~Provide credentials for legacy clients too (those that log in with username:password)~~ remember-me is not used there
- [x] Tests
- [x] Use login hook password as fallback if there's neither a device password nor a session token.
- [x] Manual testing

Next steps/PRs:
- [x] Migrate external storages to use this store too

Tested:
- External storage (webdav) using login credentials, stored in session setting and the web ui ☑️ 
- External storage (webdav) using login credentials, stored in session setting via command line using username:password ☑️ 
- External storage (webdav) using login credentials, stored in session setting via command line using username:app-password ☑️ 

There is a working POC consumer in https://github.com/nextcloud/mail/pull/28.

@LukasReschke as discussed

Fix #2732

### CODE REVIEW

**Comment by ChristophWurst** [comment]
> Provide credentials for legacy clients too (those that log in with username:password)

@LukasReschke can we expect them to send the credentials on every request? There is nothing like remember-me for clients anyway …

File: lib/private/Authentication/LoginCredentials/Credentials.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	/** @var string */
+	private $password;
+
+	/**
+	 * @param string $uid
+	 * @param string $user
+	 * @param string $password
+	 */
+	public function __construct($uid, $user, $password) {
```
Comment (nickvergessen) [inline lib/private/Authentication/LoginCredentials/Credentials.php]: uid + user is confusing, isuser displayname? then call it so.

**Review by nickvergessen — DISMISSED** [review]

File: lib/private/Authentication/LoginCredentials/Credentials.php (hunk 1 above)
Comment (ChristophWurst) [inline lib/private/Authentication/LoginCredentials/Credentials.php]: it's whatever users use to log in. depends on the setup. maybe ``loginName`` fits better

File: lib/private/Server.php [hunk 2]
```diff
@@ -241,6 +243,13 @@ public function __construct($webRoot, \OC\Config $config) {
 			});
 			return $groupManager;
 		});
+		$this->registerService(Store::class, function(Server $c) {
+			$session = $c->getSession();
```
Comment (nickvergessen) [inline lib/private/Server.php]: this needs to be added to "setSession" then:
https://github.com/nextcloud/server/pull/1943/files#diff-0a830b0354b26e84f08f705b5356a05bR949

Otherwise when it was called too early once, it always has the already closed session....

Comment (ChristophWurst) [inline lib/private/Server.php]: done

**Comment by LukasReschke** [comment]
Gonna take a look at this or see if we go back with the other approach of deactivating the remember me login with the info.xml flag :)

File: apps/encryption/appinfo/info.xml [hunk 3]
```diff
@@ -18,7 +18,6 @@
 		<user>user-encryption</user>
 		<admin>admin-encryption</admin>
 	</documentation>
-	<rememberlogin>false</rememberlogin>
```
Comment (ChristophWurst) [inline apps/encryption/appinfo/info.xml]: cc @nextcloud/encryption

Comment (rullzer) [inline apps/encryption/appinfo/info.xml]: @schiessle ^^

File: apps/encryption/appinfo/info.xml [hunk 4]
```diff
@@ -18,7 +18,6 @@
...
 	<version>1.5.0</version>
```
Comment (rullzer) [inline apps/encryption/appinfo/info.xml]: Probabaly good to give them a minor version bump?

File: apps/files_external/appinfo/info.xml [hunk 5]
```diff
@@ -12,7 +12,6 @@
 	<documentation>
 		<admin>admin-external-storage</admin>
 	</documentation>
-	<rememberlogin>false</rememberlogin>
 	<version>1.2.0</version>
```
Comment (rullzer) [inline apps/files_external/appinfo/info.xml]: Same here

**Comment by rullzer** [comment]
I rebased

**Review by rullzer — APPROVED** [review]
Awesome stuff. I did not yet manage to break it but then again my external/encryption setups are not that exciting.

Lets get this in!

@icewind1991 @schiessle have a look please!

**Comment by ChristophWurst** [comment]
@rullzer did you somehow remove some of my commits? E.g. I don't see the changes for external storages any more? 😕

**Comment by rullzer** [comment]
I hope not. Let me try again later...

**Comment by rullzer** [comment]
Shit seems I messed up. @ChristophWurst talk to me on IRC see if we can recover it on your git.

File: apps/encryption/appinfo/info.xml (hunk 4 above)
Comment (ChristophWurst) [inline apps/encryption/appinfo/info.xml]: done

File: apps/files_external/appinfo/info.xml (hunk 5 above)
Comment (ChristophWurst) [inline apps/files_external/appinfo/info.xml]: done

**Comment by ChristophWurst** [comment]
@icewind1991 @LukasReschke @nickvergessen @rullzer please review :-)

**Review by rullzer — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍
